Move PushRemoteElement to Rust

tomdale opened the pull request:
To move PushRemoteElement into Rust, this PR primarily introduces two new abstractions.

1. To allow wasm to implement the existing logic around optimizing `ConstReference`s, we extend the GBox bitfield to add an additional bit for references that have a `CONSTANT_TAG` tag.
2. We add an `UpdateWithReference` instruction to `InstructionList` that takes a reference as the first op and generates a `ReferenceCache` as well as emitting an updating program opcode that asserts on the cached reference.

alexcrichton commented on packages/@glimmer/runtime/lib/vm/gbox.ts:
```diff
@@ -130,12 +143,12 @@ export class Context {
     const idx = this.stack.length;
     this.stack.push(a);
-    return this.encodeNumberAndTag(idx, Tag.ANY);
+    return this.encodeNumberAndTag(idx, Tag.ANY, isConst(a));
```
Oh here I think you'll want to run through the generic `encode` method again with `a.value()`, right?

tomdale committed:
Add GBox encoding/decoding tests

tomdale committed:
Add Rust build artifacts to gitignore

tomdale committed:
Store ConstReference bit in GBox

tomdale committed:
Port PushRemoteElement to Rust

This relies on adding a new instruction, UpdateWithReference, to the instruction list.
